pkg/operator/operator.go

```diff
@@ -85,46 +85,43 @@ func NewOpenShiftControllerManagerOperator(
 }
 
 func (c OpenShiftControllerManagerOperator) sync() error {
-	operatorConfig, err := c.operatorConfigClient.OpenShiftControllerManagers().Get(context.TODO(), "cluster", metav1.GetOptions{})
+	cm := c.operatorConfigClient.OpenShiftControllerManagers()
+	operatorConfig, err := cm.Get(context.TODO(), "cluster", metav1.GetOptions{})
 	if err != nil {
 		return err
 	}
-	// manage status
+
 	originalOperatorConfig := operatorConfig.DeepCopy()
-	reasonString := ""
-	messageString := ""
 	switch operatorConfig.Spec.ManagementState {
 	case operatorapiv1.Removed:
+		// when removed, the same behavior of "unmanaged" will take place
 		fallthrough
-		// we equally do not allow removed/unmanaged
 	case operatorapiv1.Unmanaged:
-		reasonString = fmt.Sprintf("%sUnsupported", string(operatorConfig.Spec.ManagementState))
-		messageString = fmt.Sprintf("the controller manager spec was set to %s state, but that is unsupported, and has no effect on this condition", string(operatorConfig.Spec.ManagementState))
-
-		// as we are ignoring / not supporting unmanaged/removed, we still process any other inputs to the sync/reconciliation
-		// unlike what we used to do
+		// when unmanaged, all existing conditions will reflect the unsupported
+		// status and no additional changes on resources will be performed.
+		for _, condition := range operatorConfig.Status.Conditions {
+			condition.Reason = fmt.Sprintf("%sUnsupported", string(operatorConfig.Spec.ManagementState))
+			condition.Message = fmt.Sprintf("the controller manager spec was set to %s state, but that is unsupported, and has no effect on this condition", string(operatorConfig.Spec.ManagementState))
+			v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions, condition)
+		}
+		if !equality.Semantic.DeepEqual(operatorConfig.Status, originalOperatorConfig.Status) {
+			if _, err = cm.UpdateStatus(context.TODO(), operatorConfig, metav1.UpdateOptions{}); err != nil {
+				return err
+			}
+		}
 	case operatorapiv1.Managed:
-		// we want to empty out the reason/message string if transitioning from the other phases so the default setting above is good
-	}
-
-	for _, condition := range operatorConfig.Status.Conditions {
-		// do not change the current status as part of noting that unmanaged/removed are not supported
-		condition.Reason = reasonString
-		condition.Message = messageString
-		v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions, condition)
-	}
-	if !equality.Semantic.DeepEqual(operatorConfig.Status, originalOperatorConfig.Status) {
-		if _, err := c.operatorConfigClient.OpenShiftControllerManagers().UpdateStatus(context.TODO(), operatorConfig, metav1.UpdateOptions{}); err != nil {
-			return err
+		// when managed, all resources under control of this operator will be
+		// modified to reflect desired state.
+		forceRequeue, err := syncOpenShiftControllerManager_v311_00_to_latest(c, operatorConfig)
```

Comment:
I'm using the following snippet to investigate the conditions, and how they are changing depending on the management state.

```sh
oc get openshiftcontrollermanagers.operator.openshift.io cluster \
  --namespace="openshift-controller-manager-operator" \
  --output=json \
    |jq '.status.conditions[]'
```
### Unmanaged State

When the configuration is set to `unmanaged` state, the conditions are rewritten as:

```json
{
  "lastTransitionTime": "2020-07-12T05:13:37Z",
  "status": "False",
  "type": "ConfigObservationDegraded"
}
{
  "lastTransitionTime": "2020-08-01T08:42:44Z",
  "message": "the controller manager spec was set to Unmanaged state, but that is unsupported, and has no effect on this condition",
  "reason": "UnmanagedUnsupported",
  "status": "True",
  "type": "Available"
}
{
  "lastTransitionTime": "2020-08-01T09:04:15Z",
  "message": "the controller manager spec was set to Unmanaged state, but that is unsupported, and has no effect on this condition",
  "reason": "UnmanagedUnsupported",
  "status": "False",
  "type": "Progressing"
}
{
  "lastTransitionTime": "2020-07-12T05:13:54Z",
  "message": "the controller manager spec was set to Unmanaged state, but that is unsupported, and has no effect on this condition",
  "reason": "UnmanagedUnsupported",
  "status": "False",
  "type": "WorkloadDegraded"
}
```

### Managed State

On the default, `managed` state, the conditions are:

```json
{
  "lastTransitionTime": "2020-07-29T04:50:25Z",
  "status": "False",
  "type": "ConfigObservationDegraded"
}
{
  "lastTransitionTime": "2020-08-05T14:03:40Z",
  "status": "True",
  "type": "Available"
}
{
  "lastTransitionTime": "2020-08-05T14:03:40Z",
  "status": "False",
  "type": "Progressing"
}
{
  "lastTransitionTime": "2020-07-29T04:50:46Z",
  "status": "False",
  "type": "WorkloadDegraded"
}
```

Considering that all conditions are being mutated, @gabemontero, I think adding one more condition here, it would be lost among all others.

Reply:
I'm not following what you mean by `I think adding one more condition here, it would be lost among all others.` @otaviof 

I'm going try to re-frame things a bit ... let's go with some test-driven development:

write both a unit test and a e2e at https://github.com/openshift/cluster-openshift-controller-manager-operator/blob/master/test/e2e/cluster_openshift_controller_manager_operator_test.go that 
- validates going managed, unmanaged, and then managed again, where the conditions are checked at each stage to confirm what you see above (I bet the unit test can take care of this one)
- validates the problem noted in the bug, that the daemonset does not continually rollout when either unmanaged or managed (this one most likely can only be validated in the e2e test with an actual operator running)

whatever code change makes those new tests work will be what we go with, minus any stylistic discussion that might ensue